feat(components): Refine styling on Combobox

taylorvnoj opened the pull request:
<!--
  Atlantis uses Conventional Commits to track versions.
  Pull request titles should follow the following format.

  For help creating your pull request, you can [use this tool](https://atlantis.getjobber.com/?path=/story/guides-pull-request-title-generator--page)

  <TYPE>(<optional SCOPE>): <conditionally BREAKING CHANGE:> <description>

  eg.
    fix(SCOPE): stop graphite breaking when too much pressure applied — Patch Release
    feat(SCOPE): add 'graphiteWidth' option — (Minor) Feature Release
    feat(SCOPE): BREAKING CHANGE: remove graphiteWidth option — (Major) Breaking Release

  TYPE should consist of:
    - fix: a commit of the type fix patches a bug in your codebase
    - feat: a commit of the type feat introduces a new feature to the codebase
    - docs: documentation only changes
    - build: improvements to the build system
    - refactor: a change that neither fixes a bug nor introduces a feature
    - chore: other changes that don't modify src or test files

  SCOPE should be one of:
    - components
    - components-native
    - design
    - eslint
    - generators
    - hooks
    - stylelint


  If your pull request introduces a breaking change please append `BREAKING CHANGE:` following type / scope.

  Further Reading:
    - https://www.conventionalcommits.org
    - https://github.com/commitizen/conventional-commit-types/blob/master/index.json
-->

## Motivations

<!-- Why did you do what you did? -->
This PR is to refine and finalize styling on the Combobox component.

## Changes

<!-- https://keepachangelog.com/en/1.0.0/ -->

### Added

<!-- new features -->

Action & Options:
- pointer cursor on hover
- focus state 
- `var(--space-smaller)` gap between each
- `checkmark` icon on selected option

Filter & zero index state messges:
- proper text color and styling

### Fixed

<!-- for any bug fixes -->
Search:
- long text in the search field no longer goes behind the `clearSearch` button


## Testing

<!-- How to test your changes. -->
Find strange happenings, discrepancies & try to break things!

**Note:** fade on scroll after search and before actions, is being pushed to another ticket

---

[In Atlantis we use Github's built in pull request reviews](https://help.github.com/en/articles/about-pull-request-reviews).

![Random photo of Atlantis](https://loremflickr.com/672/400/atlantis)

chris-at-jobber commented on packages/components/src/Combobox/components/ComboboxContent/ComboboxContent.css:
```diff
@@ -21,16 +22,24 @@
+  justify-content: space-between;
+  align-items: center;
+  padding: var(--space-small);
+  color: var(--color-blue);
```
Can we use the semantic value here? 
```suggestion
  color: var(--color-heading);
```

chris-at-jobber commented on packages/components/src/Combobox/components/ComboboxTrigger/ComboboxTriggerChip/ComboboxTriggerChip.css:
```diff
@@ -1,10 +1,11 @@
   display: inline-flex;
-  gap: var(--space-small);
   padding: var(--space-small) var(--space-base);
   border: var(--border-base) solid rgb(230, 234, 235);
```
this isn't happening in this PR but where is this random RGB value coming from? Can we use one of our system colors?

chris-at-jobber replied:
Looks like `color-border` would work here

taylorvnoj replied:
I feel like I saw this rgb value in the figma at some point but `color-border` does the trick!

chris-at-jobber commented on packages/components/src/Combobox/components/ComboboxTrigger/ComboboxTriggerChip/ComboboxTriggerChip.tsx:
```diff
@@ -17,7 +17,7 @@ export function ComboboxTriggerChip(
       }}
     >
-      <Typography fontWeight="semiBold" textColor="heading">
+      <Typography fontWeight="semiBold" textColor="blue">
```
This color shouldn't be changed from the semantic to a base value

chris-at-jobber reviewed (commented):
Is it possible to adjust the height of the available scrollable area for options to make overflow more apparent?

**Current - no indication that there are more items**
![image](https://github.com/GetJobber/atlantis/assets/39704901/74fe4eaa-b05e-4741-8f76-585dbfde1142)

**Edited - more items is hinted by fader overlapping last item**
![image](https://github.com/GetJobber/atlantis/assets/39704901/2afc3832-5241-452f-afd1-7e8da418fb27)

chris-at-jobber reviewed (changes requested):
Few small changes, looking really good!

taylorvnoj commented:
Thanks @chris-at-jobber! I made those changes here 3740827348b039ba7d1b38aab73f54a5ccc69149 
Let's wait for the last of the PRs to be merged before we do another review 👍

chris-at-jobber commented:
Thanks for the updates @taylorvnoj ! Two remaining issues:

1️⃣ Looks like the options are a bit shorter than expected. In the Figma they're equivalent to 44px high, here they're:

- 37px high unselected
- 40px high selected (probably because the icon bumps them up) 

![image](https://github.com/GetJobber/atlantis/assets/39704901/9f5bb798-42eb-472e-a7cf-78a028c459cd)

Whether you want to address that here, or leave for @ZakaryH as he'll have some CSS wrangling around options to do on his end anyways, I'll leave to you two to sort.

2️⃣ The actions don't have a hover or focus state. They should have the same hover and focus state as a [tertiary button](https://atlantis.getjobber.com/?path=/story/components-actions-button-web--basic&args=type:tertiary), including having a `cursor: pointer` when hovered.

chris-at-jobber commented:
@taylorvnoj if you set it as a min-height with some padding baked into it you should get the long-label-wrapping starting from the desired minimum. In this example, I used the `calc` on the `min-height`, left the padding as-is, and set `box-sizing: border-box`:

![image](https://github.com/GetJobber/atlantis/assets/39704901/449974ff-0ba3-402e-ba61-c7048f58943a)

taylorvnoj commented:
Hey @chris-at-jobber - couple of things 😄 

017b619c3df1709cfdc4beb8feac6a90fd51a72d - commit with changes to height of the options. I also adjusted the height of the actions since the figma calls for a `40px` height.   Because of these changes I changed the height of the content once again to show more of the last option in the TriggerButton example, to show that the list continues.

4b7773547bd7f645b66c8b0145c307351c3b2d3a - hover and pointer cursor on actions.  I spoke with Laura about the focus state on Actions and the conclusion was that focus on actions will not have the yellow glow.  Do you want to loop her in and revisit this?

chris-at-jobber commented:
Changes are mostly good!

Do you have a reference for the `40px` height? In the Figma I'm referencing, the options are `44px` high:
![image](https://github.com/GetJobber/atlantis/assets/39704901/02dc3a71-ec3b-4854-b113-d3c1cdf9eafb)

I don't have context on the decision to drop the focus state from the actions, so we can revisit that later 👍

taylorvnoj commented:
@chris-at-jobber Options are 44px, Actions are 40px - both changes were done in that second to last commit

![Screenshot 2023-09-22 at 1 41 51 PM](https://github.com/GetJobber/atlantis/assets/104797704/6526bb83-b51a-4264-811d-33dfca91334f)

chris-at-jobber commented:
Ohhhh that's my bad, I conflated `actions` and `options` 😅 carry on! The working software looks great 👌 

Also, just confirmed expected hover and focus states with Laura, they've been added to [the Figma](https://www.figma.com/file/gQNDrpJYRFAwfbP3jRO0an/DSC-%5B-Combobox-%5D?type=design&node-id=2423%3A29916&mode=design&t=POFNRyuNC5xTy5fs-1)

taylorvnoj commented:
Perfect - thanks @chris-at-jobber I'll make that change here too!

dlrussel commented on packages/components/src/Combobox/components/ComboboxContent/ComboboxContent.css:
```diff
@@ -1,76 +1,93 @@
   z-index: var(--elevation-tooltip);
   max-width: var(--popover--width);
-  max-height: 400px;
+  max-height: calc(var(--space-largest) * 8.75);
```
where are these arbitrary numbers used for these calculations coming from??

taylorvnoj replied:
Based on [this comment](https://github.com/GetJobber/atlantis/pull/1530#pullrequestreview-1636377932) we need to have a max-height that shows _some_ of the next option.  There isn't a clean way to calculate this using only tokens.

Let me know if you have an alternative for the calc. I know we talked about setting 'magic numbers' but I think in this case it's a bit different.  In our `spacing.css` file and our tokens in `foundation.css` we sometimes have to calc with hard numbers.  

tldr; multiplying by 8.75 looks the best 😆

chris-at-jobber replied:
Stylistically, maybe we could define these numbers at the top of the CSS file so it's more clear in-use? Like

`--optionsOverflowHeight: calc(var(--space-largest) * 8.75)`

and then in usage, 

`max-height: var(--optionsOverflowHeight);`

?

taylorvnoj replied:
Looks like we do that in components like `Button` where specific calculations are used multiple times over.

There are two instances in this css file where we use the same calculation twice.  I'm down to set those variables in a `root` at the top of the file - how do you feel about that @dlrussel?

taylorvnoj replied:
Diane and I talked about this over zoom. I don't want to speak for her so please @dlrussel pipe in if I get this wrong - but her concern is more so about using a hardcoded, seemingly random number in the calculation _and_ for the height.  The suggestion to use variables in a `root` won't solve for that (and I am thinking we don't need that in this file).

But we do need to have a `max-height` on the options list.  With Zak's PR now merged, a `260px` max-height looks best in the options list (final decision is with Chris).

Returning to Diane's original question, my reasoning for why the calculation and height is arbitrary is because:
- we want a slight cut off on the (fifth?) option to show the fade; afaik the easiest way to do this is through css rather than our .tsx file
- there is no single token that will get us 260px
- having `calc(var(--space-small) * 32.5)` is in my opinion, faster to calculate and easier to read than trying to divide and subtract a longer list of tokens together to somehow get to `260px`

Like I said to Diane I completely understand her hesitancy and dislike for calculations like this.  Have a think on it - if there's another way to get there let's do it

taylorvnoj commented:
Future overflow indicator (gradient) work is documented in JOB-79443.

chris-at-jobber commented on packages/components/src/Combobox/components/ComboboxAction/ComboboxAction.css:
```diff
@@ -8,15 +8,22 @@
 
 .actionButton:focus {
+  box-shadow: var(--shadow-focus);
+  border-radius: var(--radius-large);
```
Couple of really minor state things that would be nice to see addressed as a fast-follow 

- is there a reason we're manipulating the border-radius on a per-state basis? As far as I can tell we could simplify this and the hover state on line25 by setting the border-radius on `.actionButton`
- can we set both the options and actions to have a `transition: all var(--timing-base);` applied to match [Button](https://github.com/GetJobber/atlantis/blob/e46abe091488d7bacf653ac8fd7ffb73b575e8aa/packages/components/src/Button/Button.css#L26C38-L26C38)?

taylorvnoj replied:
No reason, I got too lost in the css sauce. Thank you for pointing that out!

These updates plus the calc mishap are fixed here: 2303facb98cd03322324a4c0273528c94aadba93

chris-at-jobber commented on packages/components/src/Combobox/components/ComboboxContent/ComboboxContent.css:
```diff
@@ -1,75 +1,93 @@
+  top: 50%;
+  right: var(--space-base);
+  z-index: var(--elevation-tooltip);
+  width: calc(var(--space-small) * 3);
```
Fast-follow: `var(--space-small) * 3` is the same output (24px) as `var(--space-large)` so we can do away with these calcs

taylorvnoj replied:
wow Taylor 😵‍💫

chris-at-jobber commented on packages/components/src/Combobox/components/ComboboxContent/ComboboxContent.css:
```diff
@@ -1,75 +1,93 @@
+.selectedOption,
+.option:hover,
+.option:focus {
+  border-radius: var(--radius-large);
```
Same as in the actions, 

- is there a reason not to set the border-radius on the `option`, rather than modifying it based on state?
- should have the same transition property applied

chris-at-jobber commented on packages/components/src/Combobox/components/ComboboxContent/ComboboxContent.css:
```diff
@@ -87,34 +105,34 @@
-    rgba(255, 255, 255, 1)
+  background: linear-gradient(
+    180deg,
+    rgba(255, 255, 255, 0) 0%,
```
Unrelated and not a fast-follow as it's more of a "should we do this", but can/should we codify this gradient styling? We're using it all over the place in ComboBox and DataList

chris-at-jobber replied:
Note for me: we should also have guidance around when to use the fade

taylorvnoj replied:
For anyone reading, we had a quick convo about this in Office Hours today.  We'll look to extract this in someway - Chris, would you like me to put some sort of a ticket for this consideration in the backlog?

chris-at-jobber replied:
I think that makes sense, thanks!

chris-at-jobber reviewed (commented):
Design approved! My comments/notes are all non-blocking.

dlrussel commented on packages/components/src/Combobox/components/ComboboxContent/ComboboxContent.css:
```diff
@@ -1,75 +1,94 @@
 }
 
+.clearSearch {
+  display: flex;
```
Ok so we can remove this line

dlrussel commented on packages/components/src/Combobox/components/ComboboxContent/ComboboxContent.css:
```diff
@@ -1,75 +1,94 @@
+  border-radius: var(--radius-circle);
+  background-color: var(--color-surface--background);
+  transform: translateY(-50%);
+  justify-content: center;
```
we can also remove this line and the one below

dlrussel commented on packages/components/src/Combobox/components/ComboboxContent/ComboboxContent.css:
```diff
@@ -1,75 +1,94 @@
   visibility: hidden;
 }
 
+.clearSearch {
```
Can we add  ` cursor: pointer;` to this style block 👆🏻 It was historically only used when hovering over links but is now frequently used to indicate a clickable element, such as a button!

taylorvnoj replied:
Oh great catch! Thank you!

dlrussel reviewed (approved):
A few small comments but this looks really good ✅
